lib/srv/uacc/uacc_linux.go

```diff
@@ -0,0 +1,128 @@
+// +build linux
+
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package uacc
+
+// #include <stdlib.h>
+// #include "uacc.h"
+import "C"
+
+import (
+	"encoding/binary"
+	"net"
+	"sync"
+	"unsafe"
+
+	"github.com/gravitational/trace"
+)
+
+// This module concerns itself with updating the user account database and log on nodes
+// that a client connects to with an interactive session.
+
+// Due to thread safety design in glibc we must serialize all access to the accounting database.
+var accountDb sync.Mutex
+
+// Max length of username and hostname as defined by glibc.
+var nameMaxLen = 255
+
+// Open writes a new entry to the utmp database with a tag of `USER_PROCESS`.
+// This should be called when an interactive session is started.
+//
+// `username`: Name of the user the interactive session is running under.
+// `hostname`: Name of the system the user is logged into.
+// `remoteAddrV6`: IPv6 address of the remote host.
+// `ttyName`: Name of the TTY without the `/dev/` prefix.
+func Open(username, hostname string, remote net.IP, ttyName string) error {
+	rawV6 := remote.To16()
+	groupedV6 := [4]int32{}
+	for i := range groupedV6 {
+		groupedV6[i] = int32(binary.LittleEndian.Uint32(rawV6[i*4 : (i+1)*4]))
+	}
+
+	// String parameter validation.
+	if len(username) > nameMaxLen {
+		return trace.BadParameter("username length exceeds OS limits")
+	} else if len(hostname) > nameMaxLen {
+		return trace.BadParameter("hostname length exceeds OS limits")
+	} else if len(ttyName) > (int)(C.max_len_tty_name()-1) {
+		return trace.BadParameter("tty name length exceeds OS limits")
+	}
```

Comment:
nit: no need for `else if`, you can just have a series of independent `if` statements